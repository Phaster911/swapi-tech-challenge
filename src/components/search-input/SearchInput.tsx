import { ChangeEvent, useId } from "react";
import VisuallyHidden from "../visually-hidden/VisuallyHidden";
import styled from "styled-components";
import { COLORS, controlBorderRadius, QUERIES } from "../../style-constants";

const StyledLabel = styled.label`
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: auto;
  display: flex;
  align-items: stretch;
  justify-content: stretch;

  @media ${QUERIES.tabletAndUp} {
    flex-grow: 0.6;
  }
`;

const StyledInput = styled.input`
  flex: 1;
  background-color: ${({ disabled }) =>
    disabled ? COLORS.mainBackground : COLORS.pageBackground};
  border: 2px solid ${COLORS.mainBackground};
  border-radius: ${controlBorderRadius};
  padding: 0.8rem;
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "auto")};
`;

export type SearchInputProps = {
  value: string;
  disabled: boolean;
  onChangeHandler: (value: string) => void;
};

export default function SearchInput({
  value,
  disabled,
  onChangeHandler,
}: SearchInputProps) {
  const id = useId();
  return (
    <StyledLabel htmlFor={id}>
      <VisuallyHidden>Search for Star Wars characters by name</VisuallyHidden>
      <StyledInput
        id={id}
        type="search"
        placeholder="Search by character name"
        value={value}
        name="search"
        onChange={(event: ChangeEvent<HTMLInputElement>) =>
          onChangeHandler(event.target.value)
        }
        aria-label="Search for Star Wars characters by name"
        disabled={disabled}
        data-testid="search-input-testid"
      />
    </StyledLabel>
  );
}
