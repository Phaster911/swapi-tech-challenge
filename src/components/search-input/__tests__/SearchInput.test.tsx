import { act, fireEvent, render, screen } from "@testing-library/react";
import SearchInput, { SearchInputProps } from "../SearchInput";

describe("<SearchInput />", () => {
  let props: SearchInputProps;
  let onChangeFn: ReturnType<typeof vi.fn>;
  beforeEach(() => {
    onChangeFn = vi.fn();
    props = {
      value: "",
      disabled: false,
      onChangeHandler: onChangeFn,
    };
  });
  it("renders the component", () => {
    const { container } = render(<SearchInput {...props} />);
    expect(container).toMatchSnapshot();
  });

  describe("when the input disabled is true", () => {
    beforeEach(() => {
      props = {
        value: "",
        disabled: true,
        onChangeHandler: onChangeFn,
      };
    });
    it("renders the component", () => {
      const { container } = render(<SearchInput {...props} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe("when the user changes the value", () => {
    it("onChangeHandler is called with the new value", () => {
      render(<SearchInput {...props} />);
      act(() => {
        fireEvent.change(screen.getByTestId("search-input-testid"), {
          target: {
            value: "abc",
          },
        });
      });

      expect(onChangeFn).toHaveBeenCalledWith("abc");
    });
  });
});
