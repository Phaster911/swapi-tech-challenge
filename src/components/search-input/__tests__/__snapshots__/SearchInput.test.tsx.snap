// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchInput /> > renders the component 1`] = `
.c0 {
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: auto;
  display: flex;
  align-items: stretch;
  justify-content: stretch;
}

.c1 {
  flex: 1;
  background-color: white;
  border: 2px solid #e9ecee;
  border-radius: 5px;
  padding: 0.8rem;
  cursor: auto;
}

@media (min-width: 750px) {
  .c0 {
    flex-grow: 0.6;
  }
}

<div>
  <label
    class="c0"
    for=":r0:"
  >
    <span
      style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
    >
      Search for Star Wars characters by name
    </span>
    <input
      aria-label="Search for Star Wars characters by name"
      class="c1"
      data-testid="search-input-testid"
      id=":r0:"
      name="search"
      placeholder="Search by character name"
      type="search"
      value=""
    />
  </label>
</div>
`;

exports[`<SearchInput /> > when the input disabled is true > renders the component 1`] = `
.c0 {
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: auto;
  display: flex;
  align-items: stretch;
  justify-content: stretch;
}

.c1 {
  flex: 1;
  background-color: #e9ecee;
  border: 2px solid #e9ecee;
  border-radius: 5px;
  padding: 0.8rem;
  cursor: not-allowed;
}

@media (min-width: 750px) {
  .c0 {
    flex-grow: 0.6;
  }
}

<div>
  <label
    class="c0"
    for=":r1:"
  >
    <span
      style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
    >
      Search for Star Wars characters by name
    </span>
    <input
      aria-label="Search for Star Wars characters by name"
      class="c1"
      data-testid="search-input-testid"
      disabled=""
      id=":r1:"
      name="search"
      placeholder="Search by character name"
      type="search"
      value=""
    />
  </label>
</div>
`;
